extractor/docker/docker_test.go

```diff
@@ -187,35 +206,79 @@ func TestExtractFiles(t *testing.T) {
 }
 
 func TestDockerExtractor_SaveLocalImage(t *testing.T) {
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		httpPath := r.URL.String()
-		switch {
-		case strings.Contains(httpPath, "images/get?names=fooimage"):
-			_, _ = fmt.Fprint(w, "foocontent")
-		default:
-			assert.FailNow(t, "unexpected path accessed: ", r.URL.String())
+	testCases := []struct {
+		name     string
+		cacheHit bool
+	}{
+		{
+			name: "happy path with cache miss",
+		},
+		{
+			name:     "happy path with cache hit",
+			cacheHit: true,
+		},
+	}
+
+	expectedCacheBytes := "foo"
+	for _, tc := range testCases {
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			httpPath := r.URL.String()
+			switch {
+			case strings.Contains(httpPath, "images/get?names=fooimage"):
+				_, _ = fmt.Fprint(w, expectedCacheBytes)
+			default:
+				assert.FailNow(t, "unexpected path accessed: ", r.URL.String())
+			}
+		}))
+		defer ts.Close()
+
+		c, err := client.NewClientWithOpts(client.WithHost(ts.URL))
+		assert.NoError(t, err)
+
+		// setup cache
+		s, f, err := setupCache()
+		defer func() {
+			_ = f.Close()
+			_ = os.RemoveAll(f.Name())
+		}()
+		assert.NoError(t, err)
+
+		if tc.cacheHit {
+			e, _ := zstd.NewWriter(nil, zstd.WithEncoderLevel(zstd.SpeedDefault))
+			dst := e.EncodeAll([]byte(expectedCacheBytes), nil)
+			_ = s.Set(kvtypes.SetItemInput{
+				BucketName: "imagebucket",
+				Key:        "fooimage",
+				Value:      dst,
+			})
 		}
-	}))
-	defer ts.Close()
 
-	c, err := client.NewClientWithOpts(client.WithHost(ts.URL))
-	assert.NoError(t, err)
+		de := Extractor{
+			Option: types.DockerOption{},
+			Client: c,
+			Cache:  s,
+		}
 
-	// setup cache
-	tempCacheDir, _ := ioutil.TempDir("", "TestDockerExtractor_SaveLocalImage-*")
-	defer func() {
-		_ = os.RemoveAll(tempCacheDir)
-	}()
+		r, err := de.SaveLocalImage(context.TODO(), "fooimage")
+		actualSavedTarBytes, _ := ioutil.ReadAll(r)
+		assert.Equal(t, []byte("foo"), actualSavedTarBytes[:], tc.name)
```

Comment:
I'm sorry my explanation may not be enough. Let me explain again, please. How do we determine if this value "foo" is from `httptest` or a cache? Even if a cache isn't hit, the value looks the same "foo". In short, how do we know this option is effective?
https://github.com/aquasecurity/fanal/pull/51/files#diff-4d409e1b4160e28284809dc87cfddec4R218

Reply:
I understand now. I have added a commit to differentiate the expectations https://github.com/aquasecurity/fanal/pull/51/commits/5f894c996176f0bfb2faa6f5be81338a7f179e10

Reply:
I looked into the commit, but it looks the same between cached and http response data because `expectedImageData` is used here.
https://github.com/aquasecurity/fanal/commit/5f894c996176f0bfb2faa6f5be81338a7f179e10#diff-4d409e1b4160e28284809dc87cfddec4R230

In "happy path with cache hit" test, the cache data and http response are the same value, "foofromcache". We can't know if `SaveLocalImage` uses the cache or http response. 

We need to fix those values like the following

```
			case strings.Contains(httpPath, "images/get?names=fooimage"):
				_, _ = fmt.Fprint(w, "foofromdocker")
```

```
			dst := e.EncodeAll([]byte("foofromcache), nil)
```

If `SaveLocalImage` doesn't hit the cache, it returns `foofromdocker` and the test will fail because this test expect `foofromcache`. If it returns `foofromcache` as we expect, the test should succeed. Does it make sense? Feel free to ask me if you have a question.

Reply:
Yeah I understand your point. I didn't realize I had factored out the expectation to be as the actual content being served 😞 I've fixed it here: https://github.com/aquasecurity/fanal/pull/51/commits/122ee971b01e351026ecacf41f70c101c970f99e